### ISSUE TITLE
Make decoding fail on large BigDecimal values

### ISSUE THREAD

**Issue by travisbrown** [circe/circe#1363]
## Background

It's very easy to construct `BigDecimal` values in Java or Scala for which even simple arithmetic operations can take minutes:

```scala
// Using Scala's BigDecimal API on 2.12
BigDecimal("1e-100000000") + 1

// Using Java's BigDecimal API
new java.math.BigDecimal("1e-100000000").add(java.math.BigDecimal.ONE)
```

This means that systems that parse arbitrary strings into `BigDecimal` values and then do math with them are vulnerable to denial-of-service attacks, and that anyone working with `BigDecimal` values needs to take care to validate their inputs and operations.

## Deserializing big numbers

There are at least a couple of approaches a JVM serialization library can take here. Most JSON libraries follow the Java and Scala standard libraries in happily parsing `BigDecimal` values even if they're really big. Circe is one of these libraries:

```scala
scala> io.circe.jawn.decode[BigDecimal]("1e-100000000")
res0: Either[io.circe.Error,BigDecimal] = Right(1E-100000000)

scala> res0.map(_ + 1)
// Runs for a long time
```

One exception is [jsoniter-scala](https://github.com/plokhotnyuk/jsoniter-scala), where the default configuration throws an exception for this case:

```scala
import com.github.plokhotnyuk.jsoniter_scala.core._
import com.github.plokhotnyuk.jsoniter_scala.macros.CodecMakerConfig
import com.github.plokhotnyuk.jsoniter_scala.macros.JsonCodecMaker.make

implicit val bigDecimalCodec: JsonValueCodec[BigDecimal] = make(CodecMakerConfig)
```
And then:

```scala
scala> val result = readFromString[BigDecimal]("1e-100000000")
com.github.plokhotnyuk.jsoniter_scala.core.JsonReaderException: value exceeds limit for scale, offset: 0x0000000b, buf:
+----------+-------------------------------------------------+------------------+
|          |  0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f | 0123456789abcdef |
+----------+-------------------------------------------------+------------------+
| 00000000 | 31 65 2d 31 30 30 30 30 30 30 30 30             | 1e-100000000     |
+----------+-------------------------------------------------+------------------+
```

The author of jsoniter-scala has frequently argued that Circe should follow this approach (for example [here](https://github.com/circe/circe/pull/1057#issuecomment-453173300) and [here](https://www.reddit.com/r/scala/comments/eo4qx9/last_chance_to_request_changes_or_try_out_the_rcs/fecxisf/) and [here](https://stackoverflow.com/a/58777171/334519)).

## Circe's current approach

My position on this question is that protecting users from big `BigDecimal` values is [outside the scope of a JSON serialization library](https://github.com/circe/circe/pull/1057#issuecomment-453176984):

> As I see it, circe's responsibility is to provide decoding that is accurate, except when that would be excessively computationally expensive, in which case it should fail in clear and predictable ways. It's not circe's responsibility to protect people from the more nightmarish parts of the Java and Scala standard libraries.

This isn't because this kind of validation isn't important, but because it's something the user needs to be conscious of anyway, since any protection a library can provide during decoding is necessarily going to be incomplete. For example, even with jsoniter-scala something like this will take minutes to compute:

```scala
readFromString[BigDecimal]("1e-1000").pow(100000) + 1
```
In my view providing partial protection by throwing exceptions during decoding is only superficially helpful, if not actually harmful, since it could create a false sense of safety about working with `BigDecimal`.

If a user really wants to perform this kind of validation during decoding, Circe's normal validation mechanisms make this reasonably easy:

```scala
scala> import io.circe.Decoder
import io.circe.Decoder

scala> implicit val safeBigDecimalDecoder: Decoder[BigDecimal] =
     |   Decoder.decodeBigDecimal.ensure(_.scale < 100000, "value exceeds limit for scale")
safeBigDecimalDecoder: io.circe.Decoder[BigDecimal] = io.circe.Decoder$$anon$5@78dd69bf

scala> io.circe.jawn.decode[BigDecimal]("1e-100000000")
res0: Either[io.circe.Error,BigDecimal] = Left(DecodingFailure(value exceeds limit for scale, List()))
```

For these reasons I'm pretty strongly opposed to changing Circe's default `BigDecimal` decoding behavior, but I'm opening this issue to gather feedback. If you have an opinion on this question, please vote (:+1:  for changing the current behavior, :-1: for leaving it as it is) or comment.

**Comment by plokhotnyuk** [comment]
@travisbrown You are right that changing behavior and defaults is a sensitive topic.

I'm happy that circe users and easily add additional validations and change the default configuration.

IMHO a good solution would be adding a security-related page in docs where users that are going to accept untrusted input can be warned about all vulnerable cases with recommended mitigations for them.

**Comment by travisbrown** [comment]
@plokhotnyuk Yes, agreed for sure on the security documentation—will work on this soon.

**Comment by hamnis** [comment]
Ref #1040

Add documentation for this
